// unhead/dist/server.mjs
import { r as renderSSRHead$1 } from './shared/unhead.B3gLwi_2.mjs';
export { c as capoTagWeight, a as createHead, b as createServerRenderer, e as escapeHtml, p as propsToString, s as ssrRenderTags, t as tagToString } from './shared/unhead.B3gLwi_2.mjs';
import { parseHtmlForUnheadExtraction, applyHeadToHtml, parseHtmlForIndexes } from './parser.mjs';
import './shared/unhead.CfgPMHXt.mjs';
import './shared/unhead.mB5lMBMV.mjs';
import 'hookable';
import './shared/unhead.DzSj5qjO.mjs';
import './shared/unhead.fg-0ge_u.mjs';

// @__NO_SIDE_EFFECTS__
function transformHtmlTemplate(head, html) {
  const template = parseHtmlForUnheadExtraction(html);
  head.push(template.input, { _index: 0 });
  return applyHeadToHtml(template, head.render());
}
// @__NO_SIDE_EFFECTS__
function transformHtmlTemplateRaw(head, html) {
  const template = parseHtmlForIndexes(html);
  return applyHeadToHtml(template, head.render());
}

function renderSSRHead(head, options) {
  return renderSSRHead$1(head, options);
}

export { renderSSRHead, transformHtmlTemplate, transformHtmlTemplateRaw };
